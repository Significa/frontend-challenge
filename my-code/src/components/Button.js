import React from 'react';

//css
import {
	ButtonPrimaryStyle,
	ButtonPrimaryActiveStyle,
	ButtonLinkStyle,
	ButtonCardStyle
} from '../Styles/styles';


////
export const ButtonPrimary = (props) => {
	if (props.isActive) {
		return (
			<ButtonPrimaryActiveStyle type='button' onClick={props.onClick} data-testid='buttonPrimaryTrue'>
				{props.icon && <i className={props.icon} aria-hidden='true' />}
				{props.children}
			</ButtonPrimaryActiveStyle>
		);
	} else {
		return (
			<ButtonPrimaryStyle type='button' onClick={props.onClick} data-testid='buttonPrimaryFalse'>
				{props.icon && <i className={props.icon} aria-hidden='true' />}
				{props.children}
			</ButtonPrimaryStyle>
		);
	}
};

export const ButtonLink = (props) => (
	<ButtonLinkStyle type='button' onClick={props.onClick} data-testid='buttonLink'>
		{props.icon && <i className={props.icon} aria-hidden='true' />}
		{props.children}
	</ButtonLinkStyle>
);

export const ButtonCard = (props) => (
	<ButtonCardStyle type='button' onClick={props.onClick} data-testid='buttonCard'>
		{props.icon && <i className={props.icon} aria-hidden='true' data-testid='buttonCardIcon' />}
		{props.children}
	</ButtonCardStyle>
);
