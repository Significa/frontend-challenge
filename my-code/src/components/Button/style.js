import sc, { css } from 'styled-components'
import { theme } from 'styled-tools'

import { text4 } from '../../styles/text'

export const Wrapper = sc.button`
  display: flex;
  align-items: center;
  /* Account for border */
  padding: 0 15px 0 11px;
  height: 40px;
  outline: none;
  background: none;
  color: ${theme('greyLight')};
  border: 1px solid ${theme('greyMid')};
  border-radius: 4px;
  cursor: pointer;
  transition-timing-function: ${theme('transitionTiming')};
  transition-duration: ${theme('transitionDuration')};
  transition-property: color, border-color, background-color;
  font-family: inherit;
  ${text4}
  font-weight: 500;

  &:focus,
  &:hover {
    border-color: ${theme('mainRed')};
    color: ${theme('greyWhite')};
  }

  ${({ active }) =>
    active &&
    css`
      border-color: ${theme('mainRed')};
      background-color: ${theme('mainRed')};
      color: ${theme('greyWhite')};
    `}
`

export const IconWrapper = sc.div`
  width: 16px;
  height: 16px;
  margin-right: 12px;
  display: flex;
  align-items: center;
`
