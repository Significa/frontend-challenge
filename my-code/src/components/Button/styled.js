// @ flow
import styled from "styled-components"

const RegularButton = styled.button`
  align-items: center;
  display: flex;
  flex-direction: row;
  justify-content: center;

  background: none;
  border: 1px solid ${p => p.theme.color.grey.lightgrey};
  border-radius: 4px;
  color: ${p => p.theme.color.grey.lightgrey};
  font-family: ${p => p.theme.font.family};
  font-size: 1em;
  font-weight: bold;
  line-height: 19px;
  letter-spacing: 0.17;
  padding: 12px;
  outline: none;

  transition: ${p => p.theme.transition};

  &:hover {
    border: 1px solid ${p => p.theme.color.main.red};
    color: ${p => p.theme.color.grey.white};
    outline: none;
  }

  &:active {
    background-color: ${p => p.theme.color.main.red};
    outline: none;
  }
`
export default RegularButton
