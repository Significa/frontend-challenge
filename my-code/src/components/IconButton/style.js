import sc from 'styled-components'
import { theme } from 'styled-tools'

// eslint-disable-next-line import/prefer-default-export
export const Wrapper = sc.button`
  height: 40px;
  outline: none;
  border: none;
  padding: 0;
  background: none;
  color: ${theme('greyLight')};
  cursor: pointer;
  transition:
    color
    ${theme('transitionTiming')}
    ${theme('transitionDuration')}
  ;
  display: flex;
  align-items: center;
  justify-content: center;

  &:focus,
  &:hover {
    color: ${theme('greyWhite')};
  }
`
