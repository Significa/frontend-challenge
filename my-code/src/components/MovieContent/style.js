import sc from 'styled-components'
import { theme } from 'styled-tools'

import { text1, text3, text4 } from '../../styles/text'

export const Wrapper = sc.div`
  margin-top: 24px;
  display: grid;
  grid-gap: 24px 96px;
  grid-template-columns: 0.6fr 0.4fr;

  @media (max-width: 1024px) {
    grid-column-gap: 24px;
    grid-template-columns: auto minmax(30vw, 0.4fr);
  }

  @media (max-width: 700px) {
    grid-template-columns: 1fr;
  }
`

export const Row = sc.div`
  grid-column: 1;
  display: flex;
  flex-wrap: wrap;
  flex-direction: ${({ direction }) => direction || 'row'};
`

export const MetaText = sc.span`
  ${text3}
  font-weight: 400;
  color: ${theme('greyLight')};
`

export const Title = sc.span`
  ${text1}
  display: block;
  max-width: 20ch;
  margin: 0 0 8px;

  @media (max-width: 1024px) {
    font-size: 56px;
  }

  @media (max-width: 800px) {
    font-size: 48px;
  }
`

export const Text = sc.span`
  ${text4}
`

export const Heading = sc.span`
  display: block;
  margin-bottom: 8px;
  ${text4}
  font-weight: 500;
  color: ${theme('greyLight')};

  margin-top: 16px;
`

export const PosterWrapper = sc.div`
  grid-column: 2;
  grid-row: 2 / 7;

  position: relative;

  @media (max-width: 1000px) {
    grid-row: 2 / 4;
  }

  @media (max-width: 700px) {
    display: none;
  }
`

export const Poster = sc.img`
  border-radius: 8px;

  position: absolute;
  top: 0;
  right: 0;
  max-width: 100%;
`

export const Paragraph = sc.p`
  max-width: 54ch;
  margin: 0;
`

export const Subgrid = sc.div`
  grid-column: 1;
  display: grid;
  grid-gap: 56px;
  grid-template-columns: repeat(3, minmax(0, max-content));

  @media (max-width: 480px) {
    grid-template-columns: none;
    grid-gap: 24px;
  }
`
