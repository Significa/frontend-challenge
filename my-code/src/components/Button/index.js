import React from 'react'
import PropTypes from 'prop-types'

import { Wrapper, IconWrapper } from './style'

export default function Button({ icon: Icon, children, onClick, active }) {
  return (
    <Wrapper onClick={onClick} active={active}>
      {Icon && (
        <IconWrapper>
          <Icon />
        </IconWrapper>
      )}
      {children}
    </Wrapper>
  )
}

Button.propTypes = {
  icon: PropTypes.func,
  children: PropTypes.node.isRequired,
  onClick: PropTypes.func.isRequired,
  active: PropTypes.bool
}

Button.defaultProps = {
  icon: undefined,
  active: false
}
