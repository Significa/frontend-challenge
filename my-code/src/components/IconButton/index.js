import React from 'react'
import PropTypes from 'prop-types'

import { Wrapper } from './style'

export default function IconButton({ label, icon: Icon, ...rest }) {
  return (
    <Wrapper aria-label={label} {...rest}>
      <Icon />
    </Wrapper>
  )
}

IconButton.propTypes = {
  icon: PropTypes.func.isRequired,
  onClick: PropTypes.func.isRequired,
  label: PropTypes.string
}

IconButton.defaultProps = {
  label: undefined
}
