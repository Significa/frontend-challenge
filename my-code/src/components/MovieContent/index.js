import React from 'react'
import PropTypes from 'prop-types'

import Button from '../Button'
import IconButton from '../IconButton'
import IconLabel from '../IconLabel'
import TextLabel from '../TextLabel'

import Arrow from '../../assets/icon-arrow.svg'
import Heart from '../../assets/icon-heart.svg'
import HeartFilled from '../../assets/icon-heart-filled.svg'
import IMDB from '../../assets/logo-imdb.svg'
import Rotten from '../../assets/logo-rotten-tomatoes.svg'

import {
  Wrapper,
  Row,
  MetaText,
  Title,
  Text,
  Heading,
  PosterWrapper,
  Poster,
  Paragraph,
  Subgrid
} from './style'

export default function MovieContent({
  id,
  runtime,
  year,
  rated,
  title,
  imdbRating,
  rottenRating,
  favorite,
  plot,
  actors,
  genres,
  directors,
  poster,
  goBack,
  toggleFavorite
}) {
  return (
    <Wrapper>
      <Row>
        <IconButton icon={Arrow} onClick={goBack} />
      </Row>

      <Row>
        <MetaText>
          {runtime} · {year} ·
        </MetaText>
        <TextLabel>{rated}</TextLabel>
      </Row>

      <Row>
        <Title>{title}</Title>
      </Row>

      <Row>
        <IconLabel color="#ff9f1c" icon={IMDB}>
          {imdbRating}
          /10
        </IconLabel>
        {rottenRating && (
          <IconLabel color="#ff4040" icon={Rotten}>
            {rottenRating}
          </IconLabel>
        )}
        <Button
          icon={favorite ? HeartFilled : Heart}
          onClick={() => toggleFavorite(id, !favorite)}
          active={favorite}
        >
          {favorite ? 'Added' : 'Add to favourites'}
        </Button>
      </Row>

      <Row direction="column">
        <Heading>Plot</Heading>
        <Paragraph>
          <Text>{plot}</Text>
        </Paragraph>
      </Row>

      <Subgrid>
        <div>
          <Heading>Cast</Heading>
          <Text>
            {actors.map(actor => (
              <div key={actor}>{actor}</div>
            ))}
          </Text>
        </div>
        <div>
          <Heading>Genre</Heading>
          <Text>
            {genres.map(genre => (
              <div key={genre}>{genre}</div>
            ))}
          </Text>
        </div>
        <div>
          <Heading>Director</Heading>
          <Text>
            {directors.map(director => (
              <div key={director}>{director}</div>
            ))}
          </Text>
        </div>
      </Subgrid>

      <PosterWrapper>{poster && <Poster src={poster} />}</PosterWrapper>
    </Wrapper>
  )
}

MovieContent.propTypes = {
  id: PropTypes.string.isRequired,
  runtime: PropTypes.string.isRequired,
  year: PropTypes.string.isRequired,
  rated: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  imdbRating: PropTypes.string.isRequired,
  rottenRating: PropTypes.string,
  favorite: PropTypes.bool.isRequired,
  plot: PropTypes.string.isRequired,
  actors: PropTypes.arrayOf(PropTypes.string).isRequired,
  genres: PropTypes.arrayOf(PropTypes.string).isRequired,
  directors: PropTypes.arrayOf(PropTypes.string).isRequired,
  poster: PropTypes.string,
  goBack: PropTypes.func.isRequired,
  toggleFavorite: PropTypes.func.isRequired
}

MovieContent.defaultProps = {
  rottenRating: undefined,
  poster: undefined
}
