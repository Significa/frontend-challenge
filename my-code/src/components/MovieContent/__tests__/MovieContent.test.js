import React from 'react'
import { shallow } from 'enzyme'

import MovieContent from '../'

import Button from '../../Button'
import IconButton from '../../IconButton'

describe('MovieContent', () => {
  const defaultData = {
    id: 'tt0082096',
    runtime: '99 min',
    year: '2000',
    rated: 'R',
    title: 'Another Movie Title',
    imdbRating: '8.3',
    favorite: false,
    plot: 'Lorem ipsum dolor sit amet',
    actors: ['aaaa', 'bbbbbb'],
    genres: ['xxx', 'vvvvvv'],
    directors: ['ccccccc', 'ooo'],
    goBack: () => {},
    toggleFavorite: () => {}
  }

  it('renders correctly', () => {
    const tree = shallow(<MovieContent {...defaultData} />)
    expect(tree).toMatchSnapshot()
  })

  it('renders favorite correctly', () => {
    const tree = shallow(<MovieContent {...defaultData} favorite />)
    expect(tree).toMatchSnapshot()
  })

  it('renders poster correctly', () => {
    const tree = shallow(<MovieContent {...defaultData} poster="url.jpg" />)
    expect(tree).toMatchSnapshot()
  })

  it('renders rottenRating correctly', () => {
    const tree = shallow(<MovieContent {...defaultData} rottenRating="77%" />)
    expect(tree).toMatchSnapshot()
  })

  it('calls goBack', () => {
    const mockCallback = jest.fn()
    const tree = shallow(
      <MovieContent {...defaultData} goBack={mockCallback} />
    )
    tree.find(IconButton).simulate('click')
    expect(mockCallback.mock.calls.length).toBe(1)
  })

  it('calls toggleFavorite', () => {
    const mockCallback = jest.fn()
    const tree = shallow(
      <MovieContent {...defaultData} toggleFavorite={mockCallback} />
    )
    tree.find(Button).simulate('click')
    expect(mockCallback.mock.calls.length).toBe(1)
    expect(mockCallback.mock.calls[0][0]).toBe('tt0082096')
    expect(mockCallback.mock.calls[0][1]).toBe(true)
  })
})
