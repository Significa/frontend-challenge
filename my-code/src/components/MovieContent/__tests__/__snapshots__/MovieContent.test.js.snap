// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieContent renders correctly 1`] = `
<Wrapper>
  <Row>
    <IconButton
      icon={[Function]}
      onClick={[Function]}
    />
  </Row>
  <Row>
    <MetaText>
      99 min
       · 
      2000
       ·
    </MetaText>
    <TextLabel>
      R
    </TextLabel>
  </Row>
  <Row>
    <Title>
      Another Movie Title
    </Title>
  </Row>
  <Row>
    <IconLabel
      color="#ff9f1c"
      icon={[Function]}
    >
      8.3
      /10
    </IconLabel>
    <Button
      active={false}
      icon={[Function]}
      onClick={[Function]}
    >
      Add to favourites
    </Button>
  </Row>
  <Row
    direction="column"
  >
    <Heading>
      Plot
    </Heading>
    <Paragraph>
      <Text>
        Lorem ipsum dolor sit amet
      </Text>
    </Paragraph>
  </Row>
  <Subgrid>
    <div>
      <Heading>
        Cast
      </Heading>
      <Text>
        <div
          key="aaaa"
        >
          aaaa
        </div>
        <div
          key="bbbbbb"
        >
          bbbbbb
        </div>
      </Text>
    </div>
    <div>
      <Heading>
        Genre
      </Heading>
      <Text>
        <div
          key="xxx"
        >
          xxx
        </div>
        <div
          key="vvvvvv"
        >
          vvvvvv
        </div>
      </Text>
    </div>
    <div>
      <Heading>
        Director
      </Heading>
      <Text>
        <div
          key="ccccccc"
        >
          ccccccc
        </div>
        <div
          key="ooo"
        >
          ooo
        </div>
      </Text>
    </div>
  </Subgrid>
  <PosterWrapper />
</Wrapper>
`;

exports[`MovieContent renders favorite correctly 1`] = `
<Wrapper>
  <Row>
    <IconButton
      icon={[Function]}
      onClick={[Function]}
    />
  </Row>
  <Row>
    <MetaText>
      99 min
       · 
      2000
       ·
    </MetaText>
    <TextLabel>
      R
    </TextLabel>
  </Row>
  <Row>
    <Title>
      Another Movie Title
    </Title>
  </Row>
  <Row>
    <IconLabel
      color="#ff9f1c"
      icon={[Function]}
    >
      8.3
      /10
    </IconLabel>
    <Button
      active={true}
      icon={[Function]}
      onClick={[Function]}
    >
      Added
    </Button>
  </Row>
  <Row
    direction="column"
  >
    <Heading>
      Plot
    </Heading>
    <Paragraph>
      <Text>
        Lorem ipsum dolor sit amet
      </Text>
    </Paragraph>
  </Row>
  <Subgrid>
    <div>
      <Heading>
        Cast
      </Heading>
      <Text>
        <div
          key="aaaa"
        >
          aaaa
        </div>
        <div
          key="bbbbbb"
        >
          bbbbbb
        </div>
      </Text>
    </div>
    <div>
      <Heading>
        Genre
      </Heading>
      <Text>
        <div
          key="xxx"
        >
          xxx
        </div>
        <div
          key="vvvvvv"
        >
          vvvvvv
        </div>
      </Text>
    </div>
    <div>
      <Heading>
        Director
      </Heading>
      <Text>
        <div
          key="ccccccc"
        >
          ccccccc
        </div>
        <div
          key="ooo"
        >
          ooo
        </div>
      </Text>
    </div>
  </Subgrid>
  <PosterWrapper />
</Wrapper>
`;

exports[`MovieContent renders poster correctly 1`] = `
<Wrapper>
  <Row>
    <IconButton
      icon={[Function]}
      onClick={[Function]}
    />
  </Row>
  <Row>
    <MetaText>
      99 min
       · 
      2000
       ·
    </MetaText>
    <TextLabel>
      R
    </TextLabel>
  </Row>
  <Row>
    <Title>
      Another Movie Title
    </Title>
  </Row>
  <Row>
    <IconLabel
      color="#ff9f1c"
      icon={[Function]}
    >
      8.3
      /10
    </IconLabel>
    <Button
      active={false}
      icon={[Function]}
      onClick={[Function]}
    >
      Add to favourites
    </Button>
  </Row>
  <Row
    direction="column"
  >
    <Heading>
      Plot
    </Heading>
    <Paragraph>
      <Text>
        Lorem ipsum dolor sit amet
      </Text>
    </Paragraph>
  </Row>
  <Subgrid>
    <div>
      <Heading>
        Cast
      </Heading>
      <Text>
        <div
          key="aaaa"
        >
          aaaa
        </div>
        <div
          key="bbbbbb"
        >
          bbbbbb
        </div>
      </Text>
    </div>
    <div>
      <Heading>
        Genre
      </Heading>
      <Text>
        <div
          key="xxx"
        >
          xxx
        </div>
        <div
          key="vvvvvv"
        >
          vvvvvv
        </div>
      </Text>
    </div>
    <div>
      <Heading>
        Director
      </Heading>
      <Text>
        <div
          key="ccccccc"
        >
          ccccccc
        </div>
        <div
          key="ooo"
        >
          ooo
        </div>
      </Text>
    </div>
  </Subgrid>
  <PosterWrapper>
    <Poster
      src="url.jpg"
    />
  </PosterWrapper>
</Wrapper>
`;

exports[`MovieContent renders rottenRating correctly 1`] = `
<Wrapper>
  <Row>
    <IconButton
      icon={[Function]}
      onClick={[Function]}
    />
  </Row>
  <Row>
    <MetaText>
      99 min
       · 
      2000
       ·
    </MetaText>
    <TextLabel>
      R
    </TextLabel>
  </Row>
  <Row>
    <Title>
      Another Movie Title
    </Title>
  </Row>
  <Row>
    <IconLabel
      color="#ff9f1c"
      icon={[Function]}
    >
      8.3
      /10
    </IconLabel>
    <IconLabel
      color="#ff4040"
      icon={[Function]}
    >
      77%
    </IconLabel>
    <Button
      active={false}
      icon={[Function]}
      onClick={[Function]}
    >
      Add to favourites
    </Button>
  </Row>
  <Row
    direction="column"
  >
    <Heading>
      Plot
    </Heading>
    <Paragraph>
      <Text>
        Lorem ipsum dolor sit amet
      </Text>
    </Paragraph>
  </Row>
  <Subgrid>
    <div>
      <Heading>
        Cast
      </Heading>
      <Text>
        <div
          key="aaaa"
        >
          aaaa
        </div>
        <div
          key="bbbbbb"
        >
          bbbbbb
        </div>
      </Text>
    </div>
    <div>
      <Heading>
        Genre
      </Heading>
      <Text>
        <div
          key="xxx"
        >
          xxx
        </div>
        <div
          key="vvvvvv"
        >
          vvvvvv
        </div>
      </Text>
    </div>
    <div>
      <Heading>
        Director
      </Heading>
      <Text>
        <div
          key="ccccccc"
        >
          ccccccc
        </div>
        <div
          key="ooo"
        >
          ooo
        </div>
      </Text>
    </div>
  </Subgrid>
  <PosterWrapper />
</Wrapper>
`;
